### ISSUE TITLE
[torch.export] pytorch 2.7.0  torch.export failed and the error message is very confused

### ISSUE THREAD

**Issue by shykoe** [pytorch/pytorch#151094]
### 🐛 Describe the bug

here is my code
```python
import torch.nn as nn
import torch
class Smelu(nn.Module):
    def __init__(self, beta: float = 1.0):
        super(Smelu, self).__init__()
        self.beta = beta

    def forward(self, x):
        return torch.where(torch.abs(x) <= self.beta, ((x + self.beta) ** 2) / (4 * self.beta), nn.functional.relu(x))


class GateNN(nn.Module):
    def __init__(self, input_dim, hidden_unit, output_unit):
        super(GateNN, self).__init__()
        self.hidden_layer = nn.Sequential(
            nn.Linear(input_dim, hidden_unit),
            nn.ReLU()
        )
        self.output_layer = nn.Sequential(
            nn.Linear(hidden_unit, output_unit),
            nn.Sigmoid()
        )

    def forward(self, inputs, training=True):
        hidden = self.hidden_layer(inputs)
        output = 2 * self.output_layer(hidden)
        return output


class LHUC(nn.Module):
    def __init__(self, input_dim, hidden_units, gate_units):
        super(LHUC, self).__init__()
        self.hidden_units = hidden_units
        self.gate_units = gate_units
        self.gate_nn = nn.ModuleList()
        self.dense_layers = nn.ModuleList()
        self.input_dim = input_dim

        # 初始化dense layers
        dense_input = self.input_dim
        for i in range(len(self.hidden_units) - 1):
            layer = nn.Sequential(
                nn.Linear(dense_input, self.hidden_units[i]),
                Smelu()
            )
            dense_input = self.hidden_units[i]
            self.dense_layers.append(layer)
        layer = nn.Linear(self.hidden_units[-2], self.hidden_units[-1])
        self.dense_layers.append(layer)
        # 870, 400, 870
        self.gate_nn.append(GateNN(self.input_dim, self.gate_units, self.input_dim))
        input_dim = self.input_dim
        for i, unit_num in enumerate(self.hidden_units[:-1]):
            self.gate_nn.append(GateNN(self.input_dim, self.gate_units, unit_num))

    def forward(self, inputs):
        #2560, 870
        origin_embedding, auxiliary_embedding = inputs
        hidden = origin_embedding
        for i in range(len(self.hidden_units)):
            gate = self.gate_nn[i](auxiliary_embedding)
            hidden = hidden * gate
            hidden = self.dense_layers[i](hidden)
        output = hidden
        return output


batch_size = 2560


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.dht_bn = nn.BatchNorm1d(870)
        self.bias_bn = nn.BatchNorm1d(80)
        self.criterion = torch.nn.CrossEntropyLoss()
        self.dht_tower = LHUC(
            input_dim=870,
            hidden_units=[1024, 512, 256, 64, 1],
            gate_units=512)
        # [96, 64, 1]
        self.bias_mlp = nn.Sequential()
        bias_input_dim = 80
        bias_list = [96, 64, 1]
        for dim in bias_list:
            self.bias_mlp.append(nn.Linear(bias_input_dim, dim))
            self.bias_mlp.append(nn.BatchNorm1d(dim))
            self.bias_mlp.append(nn.Dropout(0.1))
            bias_input_dim = dim

    def forward(self, inputs, training=True):
        # 2560, 87, 10
        dht_table = inputs["dht_table"]
        # 2560, 870
        dht_table = dht_table.reshape([batch_size, -1])
        dht_table = self.dht_bn(dht_table)
        # 2560, 8, 10
        bias_table = inputs["bias_table"]
        bias_table = bias_table.reshape([batch_size, -1])
        bias_table = self.bias_bn(bias_table)
        features = [dht_table, dht_table]
        main_logits = self.dht_tower(features)
        main_pred = nn.functional.sigmoid(main_logits)
        bias_logits = self.bias_mlp(bias_table)
        bias_pred = nn.functional.sigmoid(bias_logits)
        pred = main_pred * bias_pred
        return {"combine_ctr_pred": pred, "dht_ctr_pred": main_pred}
    def compute_loss_and_metrics(self, labels, model_outputs, sample_weights) -> tuple[torch.Tensor, dict]:
        loss = self.criterion(model_outputs["combine_ctr_pred"].reshape(-1), labels["label"].reshape(-1))
        return loss, {"auc": [labels["label"], model_outputs["combine_ctr_pred"]], "loss": loss}
if __name__ == "__main__":
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model = Net().to(device=device)
    batch_dim = torch.export.Dim("batch", min=1, max=65536)
    example_inputs={"lhuc_table":torch.rand(2560, 15, 10).to('cuda'), "bias_table":torch.rand(2560, 8, 10).to('cuda'), "dht_table":torch.rand(2560, 87, 10).to('cuda')}
    inputs = {'inputs':{'lhuc_table':{0:batch_dim,1:torch.export.Dim.AUTO,2:torch.export.Dim.AUTO},'bias_table':{0:batch_dim,1:torch.export.Dim.AUTO,2:torch.export.Dim.AUTO}, 'dht_table':{0:batch_dim,1:torch.export.Dim.AUTO,2:torch.export.Dim.AUTO}}}
    example_tuple = tuple([example_inputs[x] for x in example_inputs.keys()])
    exported = torch.export.export(model, (example_inputs,), dynamic_shapes=inputs)
```
i got error message 
```
  - Not all values of batch = L['inputs']['lhuc_table'].size()[0] in the specified range batch <= 65536 are valid because batch was inferred to be a constant (2560).
  - Not all values of batch = L['inputs']['bias_table'].size()[0] in the specified range batch <= 65536 are valid because batch was inferred to be a constant (2560).
  - Not all values of batch = L['inputs']['dht_table'].size()[0] in the specified range batch <= 65536 are valid because batch was inferred to be a constant (2560).
 ```
the error is very confused
but when i change the dynamic_shapes to 
```
inputs = {'inputs':{'lhuc_table':{0:torch.export.Dim.AUTO,1:torch.export.Dim.AUTO,2:torch.export.Dim.AUTO},'bias_table':{0:torch.export.Dim.AUTO,1:torch.export.Dim.AUTO,2:torch.export.Dim.AUTO}, 'dht_table':{0:torch.export.Dim.AUTO,1:torch.export.Dim.AUTO,2:torch.export.Dim.AUTO}}}
exported = torch.export.export(model, (example_inputs,), dynamic_shapes=inputs)
```
it worked fine.
but I try to convert the model to aot inductor. if i use AUTO instead of  dynamic dim , i'm worried about it' performance.
  

### Versions

Collecting environment information...
PyTorch version: 2.7.0.dev20250218+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: TencentOS Linux 3.2 (Final) (x86_64)
GCC version: (GCC) 10.3.1 20210422 (Red Hat 10.3.1-1)
Clang version: 9.0.1 (Red Hat 9.0.1-2.module_el8.2.0+309+0c7b6b03)
CMake version: version 3.19.0
Libc version: glibc-2.28

Python version: 3.9.16 (main, Dec 11 2024, 20:47:20)  [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)] (64-bit runtime)
Python platform: Linux-5.4.119-1-tlinux4-0010.3-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A10
GPU 1: NVIDIA A10
GPU 2: NVIDIA A10
GPU 3: NVIDIA A10
GPU 4: NVIDIA A10
GPU 5: NVIDIA A10
GPU 6: NVIDIA A10
GPU 7: NVIDIA A10
GPU 8: NVIDIA A10
GPU 9: NVIDIA A10
GPU 10: NVIDIA A10
GPU 11: NVIDIA A10
GPU 12: NVIDIA A10
GPU 13: NVIDIA A10
GPU 14: NVIDIA A10
GPU 15: NVIDIA A10

Nvidia driver version: 525.116.04
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.9.7
/usr/lib/libcudnn_adv_infer.so.8.9.7
/usr/lib/libcudnn_adv_train.so.8.9.7
/usr/lib/libcudnn_cnn_infer.so.8.9.7
/usr/lib/libcudnn_cnn_train.so.8.9.7
/usr/lib/libcudnn_ops_infer.so.8.9.7
/usr/lib/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              208
On-line CPU(s) list: 0-207
Thread(s) per core:  2
Core(s) per socket:  26
Socket(s):           4
NUMA node(s):        4
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Platinum 8372HC CPU @ 3.40GHz
Stepping:            11
CPU MHz:             3799.983
CPU max MHz:         3401.0000
CPU min MHz:         1200.0000
BogoMIPS:            6800.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            36608K
NUMA node0 CPU(s):   0-25,104-129
NUMA node1 CPU(s):   26-51,130-155
NUMA node2 CPU(s):   52-77,156-181
NUMA node3 CPU(s):   78-103,182-207
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] nvidia-cublas-cu11==11.11.3.6
[pip3] nvidia-cuda-cupti-cu11==11.8.87
[pip3] nvidia-cuda-nvrtc-cu11==11.8.89
[pip3] nvidia-cuda-runtime-cu11==11.8.89
[pip3] nvidia-cudnn-cu11==9.1.0.70
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-curand-cu11==10.3.0.86
[pip3] nvidia-cusolver-cu11==11.4.1.48
[pip3] nvidia-cusparse-cu11==11.7.5.86
[pip3] nvidia-nccl-cu11==2.21.5
[pip3] nvidia-nvtx-cu11==11.8.86
[pip3] onnx==1.17.0
[pip3] onnxruntime==1.19.2
[pip3] onnxscript==0.2.0
[pip3] onnxsim==0.4.36
[pip3] tf2onnx==1.9.3
[pip3] torch==2.7.0.dev20250218+cu118
[pip3] triton==3.1.0
[conda] Could not collect

cc @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by angelayi** [comment]
The reason why Dim.AUTO works instead of the previous named-dim is because Dim.AUTO will not error if the dynamic dimension has been specialized to a constant value. If you print the exported program, you should see that the batch dimension is not dynamic.

As for why it gets specialized, if we replace the call to export with draft_export (on main), we can see the following guard being added:
```
1. Guard Added.
    A guard was added during tracing, which might've resulted in some incorrect
    tracing or constraint violation error.
    Specifically, this guard was added: Eq(Mod(s21*s5*s89, 2560), 0), where {'s5': "L['args'][0][0]['dht_table'].size()[0]", 's21': "L['args'][0][0]['dht_table'].size()[1]", 's89': "L['args'][0][0]['dht_table'].size()[2]"}.
    This occurred at the following stacktrace: 
        File /data/users/angelayi/pytorch/torch/nn/modules/module.py, lineno 1751, in _wrapped_call_impl
        File /data/users/angelayi/pytorch/torch/nn/modules/module.py, lineno 1762, in _call_impl
        File /data/users/angelayi/pytorch/moo.py, lineno 1609, in forward
            dht_table = dht_table.reshape([batch_size, -1]):
        
        Locals:
            dht_table: Tensor(shape: torch.Size([s5, s21, s89]), stride: (s21*s89, s89, 1), storage_offset: 0)
            batch_size: None

        Symbols:
           s5: L['args'][0][0]['dht_table'].size()[0]
           s21: L['args'][0][0]['dht_table'].size()[1]
           s89: L['args'][0][0]['dht_table'].size()[2]

    And the following framework stacktrace: 
        File /data/users/angelayi/pytorch/torch/_refs/__init__.py, lineno 4692, in view
        File /data/users/angelayi/pytorch/torch/_refs/__init__.py, lineno 3726, in _reshape_view_helper
        File /data/users/angelayi/pytorch/torch/_prims_common/__init__.py, lineno 954, in infer_size
            torch._check(

    Because of this, we have modified the dynamic shapes structure to be the
    following. You can also use torch.export.Dim.AUTO instead to specify your
    dynamic shapes, and we will automatically infer the dynamism for you.
```

Looking at the code, seems like you hard-coded the value of `batch_size = 2560`. If we modify `Net`'s forward function to instead take `batch_size` from the input, then your code works!
```python
class Net(nn.Module):
    ...
    def forward(self, inputs, training=True):
        dht_table = inputs["dht_table"]
        batch_size = dht_table.shape[0]
        ....
```

**Comment by shykoe** [comment]
@angelayi  thanks you,   it works !
